orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -328,6 +328,20 @@ void addObjectMethodsForTraining(py::module& m) {
         return py::reinterpret_steal<py::object>(dlpack::ToDlpack(v->at(idx)));
       });
 
+  py::class_<OrtValueCache>(m, "OrtValueCache")
+      .def(py::init<>())
+      .def("insert", [](OrtValueCache& cache, std::string node_arg_name, OrtValue& value) {
+        cache.CacheOrtValue(node_arg_name, value);
+      })
+      .def("keys", [](OrtValueCache& cache) {
+        std::vector<std::string> key_vec;
+        cache.GetCachedIds(key_vec);
+        return key_vec;
```

Comment:
>return key_vec; [](http://example.com/codeflow?start=8&length=15)

here you are returning a vector back to python lists I assume, can you please check the perf for this since list are opaque type and can result in a copy. 

[https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html?highlight=compilation%20unit#making-opaque-types](https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html?highlight=compilation%20unit#making-opaque-types)

"pybind11 heavily relies on a template matching mechanism to convert parameters and return values that are constructed from STL data types such as vectors, linked lists, hash tables, etc. This even works in a recursive manner, for instance to deal with lists of hash maps of pairs of elementary and custom types, etc.

However, a fundamental limitation of this approach is that internal conversions between Python and C++ types involve a copy operation that prevents pass-by-reference semantics. What does this mean?"

Reply:
Yes, this will involve two copies. First from cache to vector, and then vector will be copied to the native list. Need a way to copy direct and return py::list

Reply:
Thank you for pointing it out. I have added a TODO in the function. Since this function would be useful in debugging only, the fix can come in later.

Reply:
The fix will literally take less than hour to implement and will keep the code clean. My concern here is someone down the road can take a dependency on this call from python side while this inefficiency is present.